### PR TITLE
[IMP] project: improve various mark as done features

### PR DESCRIPTION
**Pull request by bastvdn**
This PR is a followup for the [mark as done task](https://github.com/odoo/odoo/pull/107593), aimed at fixing little bugs, improve the feature and delete useless parts.

New feature developed in this PR addresses:

- Write Javascript tests for the project.task.state widget
  - Tests are in the project_task_state_selection.js
- rewrite Research & Development demo data 
- Created a digest tip for task state selection (in Settings > technical > digest tips)
  - Tip: Use task state to keep track of the task progression
ezgif com-optimize (3)

- Fix Sharingview (Readonly) icons for the project.task.state widget, previously only the color bubble were displayed in readonly sharing views, now all of the icons are displayed
- Display state in the Calendar view
  - Display the same widget as in the kanban/list views
- Remove is_closed References
  - more details in the commit msg
- Adapt Kanban stages Exemples
  - Kanban exemples are available when creating a new stage in kanban view
image
- improve Progressbar colors (different color for approved/done, in_progress/waiting)
image

Linked PRs:
Enterprise: https://github.com/odoo/enterprise/pull/39714
Upgrade: https://github.com/odoo/upgrade/pull/4548

Task-3213526

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by bastvdn** [comment]
@odoo/rd-security 
Hello, i have a small security issue on this PR, on a markup object using almost the same tags as his neighbors :shrug: Thanks a lot

File: addons/project/models/project_project.py [hunk 1]
```diff
@@ -55,7 +59,7 @@ def _compute_task_count(self):
             for project, count in self.env['project.task'].with_context(
                 active_test=any(project.active for project in self)
             )._read_group(
-                [('is_closed', '=', False), ('project_id', 'in', self.ids)],
+                [('state', 'not in', list(CLOSED_STATES.keys())), ('project_id', 'in', self.ids)],
```
Comment (auon-odoo) [inline addons/project/models/project_project.py:62]:
```suggestion
                [('state', 'not in', list(CLOSED_STATES)), ('project_id', 'in', self.ids)],
```
This also works btw :slightly_smiling_face:

File: addons/project/models/project_project.py [hunk 2]
```diff
@@ -128,7 +132,7 @@ def _read_group_stage_ids(self, stages, domain, order):
     type_ids = fields.Many2many('project.task.type', 'project_task_type_rel', 'project_id', 'type_id', string='Tasks Stages')
     task_count = fields.Integer(compute='_compute_task_count', string="Task Count")
     task_ids = fields.One2many('project.task', 'project_id', string='Tasks',
-                               domain=[('is_closed', '=', False)])
+                               domain=[('state', 'not in', list(CLOSED_STATES.keys()))])
```
Comment (auon-odoo) [inline addons/project/models/project_project.py:135]: Same

File: addons/project/models/project_milestone.py [hunk 3]
```diff
@@ -47,10 +52,10 @@ def _compute_is_deadline_future(self):
     @api.depends('task_ids.milestone_id')
     def _compute_task_count(self):
         all_and_done_task_count_per_milestone = {
-            milestone.id: (count, is_closed_list.count(True))
+            milestone.id: (count, sum(is_closed_list.count(state) for state in CLOSED_STATES))
```
Comment (auon-odoo) [inline addons/project/models/project_milestone.py:55]:
```suggestion
            milestone.id: (count, sum(state in CLOSED_STATES for state in state_list))
```
Should call it `state_list` now.
This way sounds more pythonian, I think :neckbeard:

File: addons/project/models/project_milestone.py [hunk 4]
```diff
@@ -59,19 +64,20 @@ def _compute_task_count(self):
     def _compute_can_be_marked_as_done(self):
         if not any(self._ids):
             for milestone in self:
-                milestone.can_be_marked_as_done = not milestone.is_reached and all(milestone.task_ids.is_closed)
+                milestone.can_be_marked_as_done = not milestone.is_reached and all([True if task.state in CLOSED_STATES.keys() else False for task in milestone.task_ids])
```
Comment (auon-odoo) [inline addons/project/models/project_milestone.py:67]:
```suggestion
                milestone.can_be_marked_as_done = not milestone.is_reached and all(milestone.task_ids.mapped(lambda t: t.state in CLOSED_STATES))
```
- `True if expression else False` <=> `expression`
- Let's use `mapped`, as the framework provides it :slightly_smiling_face:
- `in dict.keys()` <=> `in dict`

File: addons/project/models/project_milestone.py [hunk 5]
```diff
@@ -59,19 +64,20 @@ def _compute_task_count(self):
...
-            ['milestone_id', 'is_closed'],
+            ['milestone_id', 'state'],
             ['__count'],
         )
         task_count_per_milestones = defaultdict(lambda: (0, 0))
-        for milestone, is_closed, count in task_read_group:
+        for milestone, state, count in task_read_group:
             opened_task_count, closed_task_count = task_count_per_milestones[milestone.id]
-            if is_closed:
+            if state in CLOSED_STATES.keys():
```
Comment (auon-odoo) [inline addons/project/models/project_milestone.py:80]:
```suggestion
            if state in CLOSED_STATES:
```

File: addons/project/static/tests/project_task_state_selection_tests.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        };
+        target = getFixture();
+        setupViewRegistries();
+    }
+}, function () {
+    QUnit.test("Check whether task state widget works as intended", async function (assert) {
+        assert.expect(4);
+
+        const views = this.views;
+        const { openView } = await start({ serverData: { views } });
```
Comment (auon-odoo) [inline addons/project/static/tests/project_task_state_selection_tests.js:51]:
```suggestion
        const { openView } = await start({ serverData: { this.views } });
```

File: addons/sale_project/models/project.py [hunk 7]
```diff
@@ -10,6 +10,11 @@
 
 from datetime import date
 
+CLOSED_STATES = {
+    '1_done': 'Done',
+    '1_canceled': 'Canceled',
+}
+
```
Comment (auon-odoo) [inline addons/sale_project/models/project.py:17]: Can't this be imported?

File: addons/sale_project/models/project.py [hunk 8]
```diff
@@ -85,7 +90,7 @@ def _compute_has_any_so_to_invoice(self):
 
     @api.depends('sale_order_id', 'task_ids.sale_order_id')
     def _compute_sale_order_count(self):
-        sale_order_items_per_project_id = self._fetch_sale_order_items_per_project_id({'project.task': [('is_closed', '=', False)]})
+        sale_order_items_per_project_id = self._fetch_sale_order_items_per_project_id({'project.task': [('state', 'not in', list(CLOSED_STATES.keys()))]})
```
Comment (auon-odoo) [inline addons/sale_project/models/project.py:93]:
```suggestion
        sale_order_items_per_project_id = self._fetch_sale_order_items_per_project_id({'project.task': [('state', 'not in', list(CLOSED_STATES))]})
```

File: addons/sale_project/models/project.py [hunk 9]
```diff
@@ -106,7 +111,7 @@ def _compute_invoice_count(self):
 
     def action_view_sos(self):
         self.ensure_one()
-        all_sale_orders = self._fetch_sale_order_items({'project.task': [('is_closed', '=', False)]}).order_id
+        all_sale_orders = self._fetch_sale_order_items({'project.task': [('state', 'not in', list(CLOSED_STATES.keys()))]}).order_id
```
Comment (auon-odoo) [inline addons/sale_project/models/project.py:114]: Same

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Thank you :cow:

File: addons/project/static/tests/project_task_state_selection_tests.js (hunk 6 above)
Comment (bastvdn) [reply addons/project/static/tests/project_task_state_selection_tests.js]: this is not allowed, check the new simplification i made :shrug:

File: addons/sale_project/models/project.py (hunk 7 above)
Comment (bastvdn) [reply addons/sale_project/models/project.py]: yup i imported this where i could now

**Review by auon-odoo — APPROVED** [review]
Lgtm, but runbot is unhappy :confused:

**Comment by mart-e** [comment]
@robodoo override=ci/security
Reason: markup on static string

**Comment by bastvdn** [comment]
@auon-odoo https://github.com/odoo/odoo/blob/b847beb46ff5c430bb852868ce291cbdcc59b8b5/addons/project/models/project_milestone.py#L7-L10
If i try the import there the runbot goes full red :shrug:

**Comment by bastvdn** [comment]
@auon-odoo Green now tho :p

File: addons/project/models/__init__.py [hunk 10]
```diff
@@ -3,11 +3,11 @@
 
 from . import account_analytic_account
 from . import mail_message
-from . import project_milestone
 from . import project_project_stage
 from . import project_task_recurrence
-# `project_task_stage_personal` has to be loaded before `project`
+# `project_task_stage_personal` has to be loaded before `project` and `project.milestone`
```
Comment (auon-odoo) [inline addons/project/models/__init__.py:8]:
```suggestion
# `project_task_stage_personal` has to be loaded before `project_project` and `project_milestone`
```

**Review by auon-odoo — APPROVED** [review]
Lgtm, thank you :kissing_heart:

File: addons/project/models/project_milestone.py [hunk 11]
```diff
@@ -2,9 +2,10 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from collections import defaultdict
-
 from odoo import api, fields, models
 
+from odoo.addons.project.models.project_task import CLOSED_STATES
```
Comment (xavierbol) [inline addons/project/models/project_milestone.py:7]: Equivalent to:
```suggestion
from .project_task import CLOSED_STATES
```
🙂

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I left some comments. 🙂 

You can also review your commits to be able to merge it. 🙂

**Review by xavierbol — APPROVED** [review]
LGTM, thanks again for your work. 🙂

**Comment by xavierbol** [comment]
robodoo r+

**Pull request closed without merging**